authorize/google_cloud_serverless.go

```diff
@@ -0,0 +1,121 @@
+package authorize
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	envoy_api_v2_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/pomerium/pomerium/authorize/evaluator"
+)
+
+var (
+	gpcIdentityTokenExpiration  = time.Hour // tokens expire after one hour according to the GCP docs
+	gcpIdentityTokenGracePeriod = time.Minute * 10
+	gcpIdentityDocURL           = "http://metadata/computeMetadata/v1/instance/service-accounts/default/identity"
+	gcpIdentityTokenSource      = NewGCPIdentityTokenSource()
+	gcpIdentityNow              = time.Now
+)
+
+// A GCPIdentityToken is an identity token for a service account.
+type GCPIdentityToken struct {
+	Audience string
+	Token    string
+	Expires  time.Time
+}
+
+// The GCPIdentityTokenSource retrieves identity tokens for service accounts. It single-flights requests and caches
+// tokens.
+type GCPIdentityTokenSource struct {
+	singleflight.Group
+
+	mu     sync.Mutex
+	tokens map[string]GCPIdentityToken
+}
+
+// NewGCPIdentityTokenSource creates a new GCPIdentityTokenSource.
+func NewGCPIdentityTokenSource() *GCPIdentityTokenSource {
+	return &GCPIdentityTokenSource{
+		tokens: make(map[string]GCPIdentityToken),
+	}
+}
+
+// Get gets an identity token.
+func (src *GCPIdentityTokenSource) Get(ctx context.Context, audience string) (string, error) {
+	v, err, _ := src.Do(audience, func() (interface{}, error) {
+		src.mu.Lock()
+		tok, ok := src.tokens[audience]
+		src.mu.Unlock()
+
+		if ok && tok.Expires.Add(-gcpIdentityTokenGracePeriod).After(gcpIdentityNow()) {
```

Comment:
Ah, I was a little fuzzy on grace period until I read this line of code. I had assumed grace period worked in the other direction.